core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -380,15 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
             if (taskState == TaskState.RUNNING) {
                 stateMachine.transitionToRunning();
             }
-            if (isFlushing()) {
+            if (isStageFlushing()) {
                 stateMachine.transitionToFlushing();
             }
-            if (finishedTasks.containsAll(allTasks)) {
+            if (isStageFinished()) {
                 stateMachine.transitionToFinished();
             }
         }
     }
 
+    private synchronized boolean isStageFlushing()
+    {
+        // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
+        return !flushingTasks.isEmpty()
+                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+    }
+
+    private synchronized boolean isStageFinished()
+    {
+        return finishedTasks.containsAll(allTasks);
+    }
+
+    private boolean addFlushingTask(TaskId taskId)
+    {
+        if (!flushingTasks.contains(taskId)) {
+            synchronized (this) {
+                return flushingTasks.add(taskId);
+            }
+        }
+        return false;
+    }
+
+    private boolean addFinishedTask(TaskId taskId)
+    {
+        boolean changed = false;
+        if (!finishedTasks.contains(taskId) || flushingTasks.contains(taskId)) {
+            synchronized (this) {
+                changed = finishedTasks.add(taskId);
+                changed |= flushingTasks.remove(taskId);
```

Comment:
no need for `OR` here. Under synchronization section task cannot be finished, but still in `flushingTasks`

Reply:
I would not say that (please look at my previous comment).